circuit/algorithms/src/bhp/hash_uncompressed.rs

```diff
@@ -44,12 +44,13 @@ impl<E: Environment, const NUM_WINDOWS: u8, const WINDOW_SIZE: u8> HashUncompres
                 // Construct the first iteration as: [ 0...0 || DOMAIN || LENGTH(INPUT) || INPUT[0..BLOCK_SIZE] ].
                 true => {
                     preimage.extend(self.domain.clone());
-                    preimage.extend(U64::constant(console::U64::new(input.len() as u64)).to_bits_le());
+                    U64::constant(console::U64::new(input.len() as u64)).write_bits_le(&mut preimage);
                     preimage.extend_from_slice(input_bits);
                 }
                 // Construct the subsequent iterations as: [ PREVIOUS_HASH[0..DATA_BITS] || INPUT[I * BLOCK_SIZE..(I + 1) * BLOCK_SIZE] ].
                 false => {
-                    preimage.extend(digest.to_x_coordinate().to_bits_le().into_iter().take(num_data_bits));
+                    digest.to_x_coordinate().write_bits_le(&mut preimage);
+                    preimage.truncate(num_data_bits);
```

Comment:
Can we leave these unchanged? I don't believe they have any major implication since they are already slice extensions.

Reply:
The use of truncation is error-prone if we add something in from in the future.

Reply:
this change actually removes an allocation, because now we no longer allocate a new vector before extending the preimage; BHP hashing is a clear "hot spot" during profiling, so it would be best if we didn't miss out on it

Reply:
would this be better?
```
false => {
    // Record the initial length of the vector in case anything is added before this point in the future.
    let initial_len = preimage.len();
    digest.to_x_coordinate().write_bits_le(&mut preimage);
    preimage.truncate(initial_len + num_data_bits);
    preimage.extend_from_slice(input_bits);
}
```

Reply:
adjusted in https://github.com/AleoHQ/snarkVM/pull/1877/commits/c3d6613747e91c7f60fd70319a3e19945adbbffe.